DEPS: upgrade Raft to 72d62a1

erikgrinaker opened the pull request:
Makes Raft leaders step down when removing themselves from the range.

```
72d62a1 2023-06-13: step down when leader removes itself [Erik Grinaker]
```

Epic: none
Release note: None

erikgrinaker committed:
kvserver: tweak `TestRaftLeaderRemovesItself`

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105578)
<!-- Reviewable:end -->

erikgrinaker commented:
For some reason, https://github.com/etcd-io/raft/pull/77 breaks leader removals. Should've known better than make opportunistic cleanups, I suppose. Not sure why it breaks yet, will dig into it tomorrow.

erikgrinaker commented:
I think the problem is that the leader steps down before we complete the joint quorum change:

```
I230627 07:36:53.597613 796 go.etcd.io/raft/v3/raft.go:1886  [T1,n1,s1,r64/1:/{Table/Max-Max}] 2883  1 switched to configuration voters=(2 3) learners=(1)
I230627 07:36:53.597678 796 go.etcd.io/raft/v3/raft.go:835  [T1,n1,s1,r64/1:/{Table/Max-Max}] 2884  1 became follower at term 6
I230627 07:36:53.597749 796 kv/kvserver/replica_raft.go:842  [T1,n1,s1,r64/1:/{Table/Max-Max},raft] 2885  raft leader changed: 1 -> 0
I230627 07:36:53.597739 2901 go.etcd.io/raft/v3/raft.go:1886  [T1,n2,s2,r64/2:/{Table/Max-Max}] 2886  2 switched to configuration voters=(2 3) learners=(1)
I230627 07:36:53.597811 3809 go.etcd.io/raft/v3/raft.go:1886  [T1,n3,s3,r64/3:/{Table/Max-Max}] 2887  3 switched to configuration voters=(2 3) learners=(1)
```

The leader is now a learner, but it needs to propose and commit the conf change that actually removes it from the range, otherwise noone else will campaign for leadership.

erikgrinaker commented:
The problem is a mismatch between Raft and CRDB. In Raft, the leader steps down once it become a learner in a joint config. In CRDB, we don't elect a new leader until the old leader is fully removed from the range -- so when the leader becomes a learner and steps down, the range remains leaderless. This patch fixes it:

```diff
--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
diff --git a/pkg/kv/kvserver/replica_raft.go b/pkg/kv/kvserver/replica_raft.go
index d33eece0989..48b1aed4727 100644
--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -2570,7 +2570,7 @@ func shouldCampaignAfterConfChange(
                // don't expect to hit this, but let's be defensive.
                return false
        }
-       if _, ok := desc.GetReplicaDescriptorByID(roachpb.ReplicaID(raftStatus.Lead)); ok {
+       if r, ok := desc.GetReplicaDescriptorByID(roachpb.ReplicaID(raftStatus.Lead)); ok && !r.IsAnyVoter() {
                // The leader is still in the descriptor.
                return false
        }
```

Unfortunately, this will break backwards compatibility with 23.1 nodes. We have a few options:

* Roll back the step-down change in etcd/raft.
* Make etcd/raft step-down opt-out or opt-in.
* Modify etcd/raft to step down on removal, not when becoming learner.

The Raft thesis describes leader removal in section 4.2.2, we should probably just do what's specified there, with an option to enable/disable it, for backwards compatibility and migrations.

erikgrinaker commented:
https://github.com/etcd-io/raft/pull/79

erikgrinaker committed:
kvserver: campaign when leader is demoted to learner

This patch campaigns when the leader demotes itself to a learner during
an atomic conf change, instead of waiting until it is completely removed
from the range. Relying on the old leader to commit the final conf
change while it's a learner appears unfortunate, and is not compatible
with an upcoming etcd/raft change that makes the learner step down in
this case.

This is backwards compatible with 23.1, because the same follower
replica is responsible for campaigning, and it does not matter if it
campaigns during the demotion or removal -- in particular because it
forces an immediate election via `forceCampaignLocked()` which
immediately bumps the term.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: enable `raft.Config.StepDownOnRemoval`

This causes the Raft leader to step down when it removes itself from the
range or demotes itself to a learner. This doesn't make any difference
functionally, since we're careful to no longer tick the replica or
otherwise use it, but in principle it could cause a leader that was no
longer part of the range to continue to assert leadership, stalling the
range since it wouldn't be able to propose anything. Stepping down
appears safer and cleaner.

Epic: none
Release note: None

erikgrinaker commented:
Closing in favor of #107203. Raft has already been bumped.
